### PR TITLE
Adds ability to add subtags and attrs after initialization of a tag

### PR DESCRIPTION
**Pull request by rashidnhm**
This pull request adds the ability to nest tags after an object has been instantiated.

This enables more dynamic generation of pages and flexibility to defer how tags are added based on different events.

The feature is fully backwards compatible.

Current approach:
```python
from rapidhtml.tags import Html, Div, H1

Html(
  Div(
    H1("Hello World!")
  ),
  Div(
    H1("Hello World!")
  )
)
```

Though the above approach can still be done, now the nested tags can be added later!

```python
from rapidhtml.tags import Html, Div, H1

root_page = Html()

hello_world_header = Div(H1("Hello World!"))

root_page.add_tag(hello_world_header)
root_page.add_tag(hello_world_header)
```

Similarly, html attributes can be added using `add_attr`

The proposed code also checks for cyclical loops.

The following code:
```python
root = Html()
div = Div()
span = Span()

div.add_tag(span)
root.add_tag(div)
span.add_tag(div)
```

Raises:
```
Traceback (most recent call last):
  File "/RapidHTML/script.py", line 15, in <module>
    span.add_tag(div)
  File "/RapidHTML/src/rapidhtml/tags.py", line 242, in add_tag
    if self._validate_cyclical_references(*tag):
  File "/RapidHTML/src/rapidhtml/tags.py", line 171, in _validate_cyclical_references
    tag._validate_cyclical_references(_parents=parents.copy())
  File "/RapidHTML/src/rapidhtml/tags.py", line 171, in _validate_cyclical_references
    tag._validate_cyclical_references(_parents=parents.copy())
  File "/RapidHTML/src/rapidhtml/tags.py", line 158, in _validate_cyclical_references
    raise custom_exceptions.CyclicalTagError(
rapidhtml.exceptions.CyclicalTagError: Cyclical reference detected! span->div->span
```

### CODE REVIEW

File: src/rapidhtml/tags.py [hunk 1]
```diff
@@ -179,6 +222,15 @@ def add_head(self, *head: "BaseTag"):
...
             self.tags.insert(0, Head(*new_head))
 
+    def add_tag(self, *tag: "BaseTag") -> None:
+        """ """
+        if self.validate_cyclical_references(*tag):
+            self.tags.extend(tag)
+
+    def add_attr(self, **attrs) -> None:
+        """ """
+        self.attrs.update(attrs)
```
Comment (ThomasJRyan) [inline src/rapidhtml/tags.py:232]: Wonder if it's worth doing some sort of check to make sure the attribute is valid. Probably more work than it's worth

File: src/rapidhtml/tags.py [hunk 2]
```diff
@@ -179,6 +222,15 @@ def add_head(self, *head: "BaseTag"):
                 new_head = head
             self.tags.insert(0, Head(*new_head))
 
+    def add_tag(self, *tag: "BaseTag") -> None:
+        """ """
```
Comment (ThomasJRyan) [inline src/rapidhtml/tags.py:226]: Docstring should be filled in

File: src/rapidhtml/tags.py [hunk 3]
```diff
@@ -179,6 +222,15 @@ def add_head(self, *head: "BaseTag"):
...
+        if self.validate_cyclical_references(*tag):
+            self.tags.extend(tag)
+
+    def add_attr(self, **attrs) -> None:
+        """ """
```
Comment (ThomasJRyan) [inline src/rapidhtml/tags.py:231]: Same

File: src/rapidhtml/tags.py [hunk 4]
```diff
@@ -121,11 +122,53 @@ def __init__(
...
 
     def __init_subclass__(cls, self_closing: bool = False, **kwargs) -> None:
         super().__init_subclass__(**kwargs)
         cls.__self_closing = self_closing
 
+    @property
+    def _uuid(self) -> str:
+        return self.__uuid
+
+    def validate_cyclical_references(
```
Comment (ThomasJRyan) [inline src/rapidhtml/tags.py:137]: Should we prepend with `_`?
```suggestion
    def _validate_cyclical_references(
```

File: src/rapidhtml/tags.py [hunk 5]
```diff
@@ -121,11 +122,53 @@ def __init__(
...
+            ValueError: If a cyclical reference is detected.
+
+        Returns:
+            bool: True if no cyclical references are found.
+        """
+
+        parents = _parents or []
+
+        if self.__uuid in parents:
+            raise ValueError("Cyclical reference detected!")
```
Comment (ThomasJRyan) [inline src/rapidhtml/tags.py:157]: I'm thinking we should get a `rapidhtml.errors` started and create a `CyclicalReferenceError` to raise here. Admittedly it doesn't serve much of a purpose other than just looking cool and getting specific with thrown errors

**Review by ThomasJRyan — CHANGES_REQUESTED** [review]
Honestly the logic seems pretty solid. Only thing I'd want more from this are unittests, including a test with an arbitrarily large number of tags. I'm curious how long it'd take to recurse through all of them

File: src/rapidhtml/tags.py (hunk 1 above)
Comment (rashidnhm) [reply src/rapidhtml/tags.py:232]: This is a very expensive check to be honest, there are lots of core html attributes and if you plan to use some JS based framework then it blows up even larger. I shall skip the check for now

File: src/rapidhtml/tags.py (hunk 2 above)
Comment (rashidnhm) [reply src/rapidhtml/tags.py]: Done

File: src/rapidhtml/tags.py (hunk 5 above)
Comment (rashidnhm) [reply src/rapidhtml/tags.py]: I've added a custom exception block. It's more stylistic for sure but a nice one to have

**Review by ThomasJRyan — APPROVED** [review]

**Pull request merged**